File: 92-square-digit-chains/square_digit_chains.py
def square_digits(n):
    result = 0
    while n:
        n, digit = divmod(n, 10)
        result += digit**2
    return result


def one_or_89(n):
    while n != 1 and n != 89:
        n = square_digits(n)
    return n


def main():
    count = 0
    for n in xrange(1, 10000000):
        if one_or_89(n) == 89:
            count += 1
    print count


if __name__ == "__main__":
    main()

File: 92-square-digit-chains/square_digit_chains.cpp
#include <iostream>

int squareDigits(int n) {
  int result = 0;
  while (n) {
    int digit = n % 10;
    result += digit * digit;
    n /= 10;
  }
  return result;
}

int oneOr89(int n) {
  while (n != 1 && n != 89) n = squareDigits(n);
  return n;
}

int main() {
  int count = 0;
  for (int n = 1; n < 10000000; n++) {
    if (oneOr89(n) == 89) count++;
  }
  std::cout << count << std::endl;

  return 0;
}
